### ISSUE TITLE
np.ceil and np.floor are inconsistent with math.ceil and math.floor

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#9068]
In Python 3, `floor` and `ceil` changed to return integers, rather than floats:

```python
>>> math.floor(1.5), math.ceil(1.5)
(1, 2)
>>> np.floor(1.5), np.ceil(1.5)
(1.0, 2.0)   # also the output for the first case on 2.7
```

Should we update these function in numpy to also return integer types?

**Comment by charris** [comment]
That's a pretty major change which we might want to put off until we drop support fot 2.7. Note that the C functions return floats, and the C functions are what numpy uses. OTOH, integers are usually what folks want and I've considered adding `ifloor` and `iceil` functions as integer versions in the past. Note that Python has an advantage here as the integer type has unlimited precision.

**Comment by endolith** [comment]
What would happen for `np.floor([1.5, 1e300])`?  Python int type can be arbitrarily long

**Comment by eric-wieser** [comment]
> What would happen for np.floor([1.5, 1e300])

The same thing as `x = np.array[1.5, 1e300])); np.int64(x)` would make sense. That should probably give a warning too.

The workaround would be to explicitly request a float, with `np.floor([1.5, 1e300], dtype=float)`

**Comment by njsmith** [comment]
> The workaround would be to explicitly request a float, with `np.floor(..., dtype=float)`

I don't think this works given how ufunc dispatch is handled. If you had a `float,float->int` loop then saying `dtype=float` would cast to int and then cast back to float, which is no use. But I think that's the only way currently to get the int-by-default behavior. You'd need like... Two versions of the loop that differ only in output type, and then some way to make sure that the integer version has higher "priority" than the float version. Or something. (Actually for all I know that's possible by carefully arranging the lookup table in the right order; the loop dispatch code is tricky. But it doesn't seem like something to rely on.)

**Comment by kalvdans** [comment]
I vote for documenting that np.ceil and np.floor returns float and be done with it. I can make a pull request if you agree.

**Comment by anntzer** [comment]
As in #5700 I vote for returning ints.  Especially now that numpy is a bit picky with requiring ints for indices (which is good IMO), having to cast the return value of float is quite annoying (and gratuitiously requires another intermediate buffer).

**Comment by Zac-HD** [comment]
There is also some surprising behaviour with the Python 3 `math.ceil` and `math.floor` functions, found via HypothesisWorks/hypothesis#1667:

```python
>>> np.dtype('int64').type(2 ** 53 + 1)
9007199254740993
>>> math.ceil(_)
9007199254740992
```

Just as not all floats are representable as integers, 64-bit integers can be unrepresentable as floats - and there's an intermediate cast in the `__ceil__` method.  I don't have a proposed solution for this and it wasn't hard to work out what's happening, but even knowing that `np.ceil` returns a float this surprised me.

**Comment by mhvk** [comment]
Found this while looking at old issues: can we avoid breaking backward compatibility by adding new `f->i` loops so that at least one can select integer output? (i.e., the opposite of the `dtype=float` suggestion above).

**Comment by charris** [comment]
Perhaps it would be better to add new functions, `iceil`, `ifloor`, etc. That floats eventually develop integer sized gaps could be a problem, as already pointed out. I think that dealing with that could be tricky if we want to do a bit better in the integer domain.

**Comment by mhvk** [comment]
@charris - yes, perhaps that is best. It could then extent to `iround`, `ifix`, etc., too

**Comment by harsharajkumar** [comment]
Hi! I’ve been looking through this issue and I agree that the current behavior can be a bit confusing, especially compared to Python’s `math.floor` and `math.ceil`.

That said, instead of changing the return type (which could introduce backward compatibility issues), it might be more practical to improve the documentation to make this behavior clearer.

For example, the docs could explicitly highlight the difference:

math.ceil(1.5)   # 2 (int)
np.ceil(1.5)     # 2.0 (float)

NumPy’s versions always return floating-point values, even for scalar inputs, since they are designed to operate consistently on arrays. It could also suggest casting when integer output is desired:

int(np.ceil(1.5))        # 2
np.ceil([1.2, 2.3]).astype(int)  # array([2, 3])

I’d be happy to help put together a documentation update along these lines if that sounds useful!
